python/cuml/dask/neighbors/nearest_neighbors.py

```diff
@@ -13,572 +13,229 @@
-        that currently hold the parts of given futures.
-        :param gpu_futures:
-        :param client:
-        :return:
-        """
-        who_has = client.who_has(gpu_futures)
+    def _func_kneighbors(model, local_idx_parts, idx_m, n, idx_partsToRanks,
+                         local_query_parts, query_m, query_partsToRanks,
+                         rank, k):
+        return model.kneighbors(
+            local_idx_parts, idx_m, n, idx_partsToRanks,
+            local_query_parts, query_m, query_partsToRanks,
+            rank, k
+        )
 
-        key_to_host_dict = {}
-        for key in who_has:
-            key_to_host_dict[key] = parse_host_port(who_has[key][0])
+    @staticmethod
+    def _func_get_d(f, idx):
+        i, d = f
+        return d[idx]
 
-        hosts_to_key_dict = {}
-        for key, host in key_to_host_dict.items():
-            if host not in hosts_to_key_dict:
-                hosts_to_key_dict[host] = set([key])
+    @staticmethod
+    def _func_get_i(f, idx):
+        i, d = f
+        return i[idx]
+
+    def kneighbors(self, X=None, n_neighbors=None, return_distance=True,
+                   _return_futures=False):
+        """
+        Query the distributed nearest neighbors index
+        :param X : dask_cudf.Dataframe Vectors to query. If not
+                   provided, neighbors of each indexed point are returned.
+        :param n_neighbors : Number of neighbors to query for each row in
+                             X. If not provided, the n_neighbors on the
+                             model are used.
+        :param return_distance : If false, only indices are returned
+        :return : dask_cudf.DataFrame containing distances
+        :return : dask_cudf.DataFrame containing indices
+        """
+        if n_neighbors is None:
+            if "n_neighbors" in self.kwargs \
+                    and self.kwargs["n_neighbors"] is not None:
+                n_neighbors = self.kwargs["n_neighbors"]
             else:
-                hosts_to_key_dict[host].add(key)
+                try:
+                    from cuml.neighbors.nearest_neighbors_mg import \
+                        NearestNeighborsMG as cumlNN
+                except ImportError:
+                    _raise_import_exception()
+                n_neighbors = cumlNN().n_neighbors
+
+        X = self.X if X is None else X
+
+        if X is None:
+            raise ValueError("Model needs to be trained using fit() "
+                             "before calling kneighbors()")
+
+        query_futures = self.client.sync(extract_ddf_partitions,
+                                         X, agg=False)
+
+        index_worker_to_parts = OrderedDict()
+        for w, p in self.index_futures:
+            if w not in index_worker_to_parts:
+                index_worker_to_parts[w] = []
+            index_worker_to_parts[w].append(p)
+
+        query_worker_to_parts = OrderedDict()
+        for w, p in query_futures:
```

Comment:
I'm building this separately from the `extract_ddf_partitions` because I needed the expanded sorted array of partitions to build the `partsToRanks` arrays. This is why the `agg=False` was used as an argument to `extract_ddf_partitions`